### PR TITLE
[GHA] Better handling of environment variables and change detection

### PR DESCRIPTION
**Pull request by heapcrash**
This should help reduce the number of times that Android AVD tests have to be run, along with consolidating some of the environment variable logic.

### CODE REVIEW

**Comment by heapcrash** [comment]
This might not work because `github.event.pull_request.base.ref` may not be defined for `push`.  @Arusekk do you have a better way to handle this?

**Comment by heapcrash** [comment]
Verified that changelog detection is skipped with the most recent commit.

**Review by Arusekk — COMMENTED** [review]
Looks good, but it will run AVD tests for direct commits on branches (`GITHUB_BASE='origin/'` is an invalid ref). Guess we will now need to commit features through branches as it should be done. :) The bad (or good) thing is, it will run also for merge commits from merged PRs.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,5 +1,8 @@
 name: Continuous Integration
 on: [push, pull_request]
+env:
+  GITHUB_BASE: origin/${{ github.event.pull_request.base.ref }}
```
Comment (Arusekk) [inline .github/workflows/ci.yml:4]: We can add another variable that would hold `github.event.before`, and use that in case the event is a sole `push`.

Comment (zachriggle) [reply .github/workflows/ci.yml:4]: `event.before` often refers to the previously-CI'ed commit AFAICT and isn't what we wanted

**Comment by zachriggle** [comment]
I don't mind if the AVD tests run in the background post-merge.  If no Android stuff changed, I find it pretty unlikely that they would have broken -- but at least we'll be able to tell

File: .github/workflows/ci.yml (hunk 1 above)
Comment (zachriggle) [reply .github/workflows/ci.yml:4]: Where can I find documentation for these fields?  It looked like `event.before` was not actually what we wanted

**Comment by Arusekk** [comment]
The `ci.yml` part got fixed in #1641, but the `changelog.yml` changes are totally cool.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (heapcrash) [reply .github/workflows/ci.yml:4]: I believe `github.event.before` is not the target branch ref, but rather the literal previous thing that was tested via GHA (e.g.the previous commit)

**Comment by heapcrash** [comment]
@Arusekk do you approve of this?

**Comment by heapcrash** [comment]
I think I just accidentally'd that merge.  DO NOT MERGE until I have a chance to look at this again

**Comment by Arusekk** [comment]
This also looks inferior to #1594, or I am missing something

**Comment by heapcrash** [comment]
Need to follow up on this one.

**Pull request closed without merging**